Regression: Transitive prebuilt_dependencies are too lenient

Currently, if `prebuilt_dependencies` like "base" are listed in a rule, than any other rule depending on it doesn't need to list them itself.

In particular, consider this test diff:
```
diff --git a/tests/library-deps/BUILD b/tests/library-deps/BUILD
index 08cd396..bc5c065 100644
--- a/tests/library-deps/BUILD
+++ b/tests/library-deps/BUILD
@@ -10,7 +10,6 @@ haskell_library(
   name = "library-deps",
   srcs = ["TestLib.hs"],
   deps = ["//tests/library-deps/sublib"],
-  prebuilt_dependencies = ["base"],
   visibility = ["//visibility:public"],
 )
 
@@ -19,7 +18,6 @@ haskell_test(
   srcs = ["Bin.hs"],
   main_file = "Bin.hs",
   deps = ["//tests/library-deps/sublib"],
-  prebuilt_dependencies = ["base"],
   visibility = ["//visibility:public"],
   size = "small",
 )
```

We used to get an error (tested on 582f2d71fd1b1d29c5a2d6ac156f0993e7e7c4af):
```
$ bazel build tests/library-deps:bin-deps
tests/library-deps/Bin.hs:1:8: error:
    Could not find module ‘Prelude’
    Use -v to see a list of the files searched for.
  |
1 | module Main (main) where
  |        ^^^^
```

But on the latest master, the command incorrectly succeeds.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -21,6 +21,9 @@ jobs:
           name: Run tests
           command: bazel test //... --config=ci
       - run:
+          name: Test targets that must fail
+          command: tests/targets-that-must-fail/run-tests.sh
+      - run:
           name: Test REPL for libraries
           command: |
             bazel build //tests/repl-targets:hs-lib-repl
@@ -67,6 +70,10 @@ jobs:
           name: Run tests
           shell: /bin/bash -eilo pipefail
           command: bazel test //... --config=ci
+      - run:
+          name: Test targets that must fail
+          shell: /bin/bash -eilo pipefail
+          command: tests/targets-that-must-fail/run-tests.sh
 
 workflows:
   version: 2

--- a/WORKSPACE
+++ b/WORKSPACE
@@ -22,6 +22,11 @@ nixpkgs_git_repository(
   revision = "7c3dc2f53fc837be79426f11c9133f73d15a05c4",
 )
 
+local_repository(
+  name = "targets_that_must_fail",
+  path = "tests/targets-that-must-fail",
+)
+
 nixpkgs_package(
   name = "ghc",
   repository = "@nixpkgs",

--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -217,13 +217,13 @@ def _compilation_defaults(hs, cc, java, dep_info, srcs, cpp_defines, compiler_fl
   haddock_args.add(ih_root_arg, before_each="--optghc")
 
   # Expose all prebuilt dependencies
-  for prebuilt_dep in set.to_list(dep_info.prebuilt_dependencies):
+  for prebuilt_dep in set.to_list(dep_info.direct_prebuilt_deps):
     items = ["-package", prebuilt_dep]
     args.add(items)
     haddock_args.add(items, before_each="--optghc")
 
   # Expose all bazel dependencies
-  for package in set.to_list(dep_info.package_ids):
+  for package in set.to_list(dep_info.direct_package_ids):
     items = ["-package-id", package]
     args.add(items)
     if package != my_pkg_id:

--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -223,7 +223,7 @@ def _compilation_defaults(hs, cc, java, dep_info, srcs, cpp_defines, compiler_fl
     haddock_args.add(items, before_each="--optghc")
 
   # Expose all bazel dependencies
-  for package in set.to_list(dep_info.direct_package_ids):
+  for package in set.to_list(dep_info.package_ids):
     items = ["-package-id", package]
     args.add(items)
     if package != my_pkg_id:

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -272,7 +272,7 @@ def link_library_dynamic(hs, dep_info, object_files, my_pkg_id):
   if hs.toolchain.is_darwin:
     args.add(["-optl-Wl,-dead_strip_dylibs"])
 
-  for package in set.to_list(dep_info.direct_package_ids):
+  for package in set.to_list(dep_info.package_ids):
     args.add(["-package-id", package])
   # XXX This should be really dep_info.direct_prebuilt_deps, but since we
   # cannot add prebuilt_dependencies to the "depends" field on package

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -272,8 +272,12 @@ def link_library_dynamic(hs, dep_info, object_files, my_pkg_id):
   if hs.toolchain.is_darwin:
     args.add(["-optl-Wl,-dead_strip_dylibs"])
 
-  for package in set.to_list(dep_info.package_ids):
+  for package in set.to_list(dep_info.direct_package_ids):
     args.add(["-package-id", package])
+  # XXX This should be really dep_info.direct_prebuilt_deps, but since we
+  # cannot add prebuilt_dependencies to the "depends" field on package
+  # registration (see a comment there), we have to pass all transitive
+  # prebuilt_dependencies on linking like this.
   for package in set.to_list(dep_info.prebuilt_dependencies):
     args.add(["-package", package])
 

--- a/haskell/private/actions/package.bzl
+++ b/haskell/private/actions/package.bzl
@@ -5,7 +5,7 @@ load(":private/pkg_id.bzl", "pkg_id")
 load(":private/set.bzl", "set")
 load("@bazel_skylib//:lib.bzl", "paths")
 
-def package(hs, dep_info, interfaces_dir, static_library, dynamic_library, exposed_modules, other_modules, my_pkg_id, pkg_deps):
+def package(hs, dep_info, interfaces_dir, static_library, dynamic_library, exposed_modules, other_modules, my_pkg_id):
   """Create GHC package using ghc-pkg.
 
   Args:
@@ -36,7 +36,11 @@ def package(hs, dep_info, interfaces_dir, static_library, dynamic_library, expos
     "dynamic-library-dirs": "${pkgroot}",
     "hs-libraries": "HS" + pkg_id.to_string(my_pkg_id),
     "depends":
-      ", ".join([d.package_id for d in pkg_deps]),
+      ", ".join(
+        # XXX Ideally we would like to specify here prebuilt dependencies
+        # too, but we don't know their versions, and package ids without
+        # versions will be rejected as unknown.
+        set.to_list(dep_info.direct_package_ids)),
   }
   hs.actions.write(
     output=registration_file,

--- a/haskell/private/actions/package.bzl
+++ b/haskell/private/actions/package.bzl
@@ -40,7 +40,7 @@ def package(hs, dep_info, interfaces_dir, static_library, dynamic_library, expos
         # XXX Ideally we would like to specify here prebuilt dependencies
         # too, but we don't know their versions, and package ids without
         # versions will be rejected as unknown.
-        set.to_list(dep_info.direct_package_ids)),
+        set.to_list(dep_info.package_ids)),
   }
   hs.actions.write(
     output=registration_file,

--- a/haskell/private/dependencies.bzl
+++ b/haskell/private/dependencies.bzl
@@ -109,6 +109,8 @@ def gather_dep_info(ctx):
     interface_files = set.empty(),
     prebuilt_dependencies = set.from_list(ctx.attr.prebuilt_dependencies),
     external_libraries = {},
+    direct_prebuilt_deps = set.from_list(ctx.attr.prebuilt_dependencies),
+    direct_package_ids = set.empty(),
   )
 
   for dep in ctx.attr.deps:
@@ -127,7 +129,12 @@ def gather_dep_info(ctx):
         dynamic_libraries = set.mutable_union(acc.dynamic_libraries, binfo.dynamic_libraries),
         interface_files = set.mutable_union(acc.interface_files, binfo.interface_files),
         prebuilt_dependencies = set.mutable_union(acc.prebuilt_dependencies, binfo.prebuilt_dependencies),
-        external_libraries = dicts.add(acc.external_libraries, binfo.external_libraries)
+        external_libraries = dicts.add(acc.external_libraries, binfo.external_libraries),
+        direct_prebuilt_deps = acc.direct_prebuilt_deps,
+        direct_package_ids = set.mutable_insert(
+          acc.direct_package_ids,
+          dep[HaskellLibraryInfo].package_id,
+        )
       )
     else:
       # If not a Haskell dependency, pass it through as-is to the
@@ -149,6 +156,8 @@ def gather_dep_info(ctx):
             for f in dep.files.to_list() if _is_shared_library(f)}
 
           ),
+        direct_prebuilt_deps = acc.direct_prebuilt_deps,
+        direct_package_ids = acc.direct_package_ids,
       )
 
   return acc

--- a/haskell/private/dependencies.bzl
+++ b/haskell/private/dependencies.bzl
@@ -110,7 +110,6 @@ def gather_dep_info(ctx):
     prebuilt_dependencies = set.from_list(ctx.attr.prebuilt_dependencies),
     external_libraries = {},
     direct_prebuilt_deps = set.from_list(ctx.attr.prebuilt_dependencies),
-    direct_package_ids = set.empty(),
   )
 
   for dep in ctx.attr.deps:
@@ -131,10 +130,6 @@ def gather_dep_info(ctx):
         prebuilt_dependencies = set.mutable_union(acc.prebuilt_dependencies, binfo.prebuilt_dependencies),
         external_libraries = dicts.add(acc.external_libraries, binfo.external_libraries),
         direct_prebuilt_deps = acc.direct_prebuilt_deps,
-        direct_package_ids = set.mutable_insert(
-          acc.direct_package_ids,
-          dep[HaskellLibraryInfo].package_id,
-        )
       )
     else:
       # If not a Haskell dependency, pass it through as-is to the
@@ -157,7 +152,6 @@ def gather_dep_info(ctx):
 
           ),
         direct_prebuilt_deps = acc.direct_prebuilt_deps,
-        direct_package_ids = acc.direct_package_ids,
       )
 
   return acc

--- a/haskell/private/haskell_impl.bzl
+++ b/haskell/private/haskell_impl.bzl
@@ -120,10 +120,6 @@ def haskell_library_impl(ctx):
     if not set.is_member(other_modules, module):
       set.mutable_insert(exposed_modules, module)
 
-  pkg_deps = [
-    d[HaskellLibraryInfo] for d in ctx.attr.deps
-    if HaskellLibraryInfo in d
-  ]
   conf_file, cache_file = package(
     hs,
     dep_info,
@@ -133,7 +129,6 @@ def haskell_library_impl(ctx):
     exposed_modules,
     other_modules,
     my_pkg_id,
-    pkg_deps,
   )
 
   build_info = HaskellBuildInfo(

--- a/haskell/private/providers.bzl
+++ b/haskell/private/providers.bzl
@@ -30,6 +30,8 @@ HaskellBuildInfo = provider(
     "interface_files": "Set of interface files belonging to the packages.",
     "prebuilt_dependencies": "Transitive collection of names of wired-in Haskell dependencies.",
     "external_libraries": "Set of dynamic shared libraries needed for linking.",
+    "direct_prebuilt_deps": "Set of direct prebuilt dependencies.",
+    "direct_package_ids": "Set of package ids of direct dependencies.",
   },
 )
 

--- a/haskell/private/providers.bzl
+++ b/haskell/private/providers.bzl
@@ -22,7 +22,7 @@ DefaultCompileInfo = provider(
 HaskellBuildInfo = provider(
   doc = "Common information about build process: dependencies, etc.",
   fields = {
-    "package_ids": "Set of all package ids of transitive dependencies.",
+    "package_ids": "Set of all package ids of direct (non-prebuilt) dependencies.",
     "package_confs": "Set of package .conf files.",
     "package_caches": "Set of package cache files.",
     "static_libraries": "Ordered collection of compiled library archives.",
@@ -31,7 +31,6 @@ HaskellBuildInfo = provider(
     "prebuilt_dependencies": "Transitive collection of names of wired-in Haskell dependencies.",
     "external_libraries": "Set of dynamic shared libraries needed for linking.",
     "direct_prebuilt_deps": "Set of direct prebuilt dependencies.",
-    "direct_package_ids": "Set of package ids of direct dependencies.",
   },
 )
 

--- a/tests/targets-that-must-fail/WORKSPACE
+++ b/tests/targets-that-must-fail/WORKSPACE
@@ -1,0 +1,1 @@
+workspace(name = "targets_that_must_fail")

--- a/tests/targets-that-must-fail/run-tests.sh
+++ b/tests/targets-that-must-fail/run-tests.sh
@@ -1,0 +1,24 @@
+#!/bin/sh
+
+set -e
+
+test_transitive_deps() {
+    set +e
+
+    bazel build @targets_that_must_fail//transitive-deps:lib-c
+    if [ $? -eq 0 ]; then
+        echo "bazel build @targets_that_must_fail//transitive-deps:lib-c should have failed."
+        exit 1
+    fi
+
+    bazel build @targets_that_must_fail//transitive-deps:lib-d
+    if [ $? -eq 0 ]; then
+        echo "bazel build @targets_that_must_fail//transitive-deps:lib-d should have failed."
+        exit 1
+    fi
+
+    set -e
+    exit 0
+}
+
+test_transitive_deps

--- a/tests/targets-that-must-fail/transitive-deps/BUILD
+++ b/tests/targets-that-must-fail/transitive-deps/BUILD
@@ -1,0 +1,35 @@
+package(default_testonly = 1)
+
+load(
+  "@io_tweag_rules_haskell//haskell:haskell.bzl",
+  "haskell_library",
+)
+
+haskell_library(
+  name = "lib-a",
+  srcs = ["LibA.hs"],
+  prebuilt_dependencies = ["base"],
+)
+
+haskell_library(
+  name = "lib-b",
+  srcs = ["LibB.hs"],
+  deps = [":lib-a"],
+  prebuilt_dependencies = ["base"],
+)
+
+haskell_library(
+  # Should fail because it doesn't specify "base" explicitly.
+  name = "lib-c",
+  srcs = ["LibC.hs"],
+  deps = [":lib-b"],
+  prebuilt_dependencies = [],
+)
+
+haskell_library(
+  # Should fail because it doesn't specify "lib-a" explicitly.
+  name = "lib-d",
+  srcs = ["LibD.hs"],
+  deps = [":lib-b"],
+  prebuilt_dependencies = ["base"],
+)

--- a/tests/targets-that-must-fail/transitive-deps/LibA.hs
+++ b/tests/targets-that-must-fail/transitive-deps/LibA.hs
@@ -1,0 +1,4 @@
+module LibA (thingA) where
+
+thingA :: Int
+thingA = 5

--- a/tests/targets-that-must-fail/transitive-deps/LibB.hs
+++ b/tests/targets-that-must-fail/transitive-deps/LibB.hs
@@ -1,0 +1,6 @@
+module LibB (thingB) where
+
+import LibA (thingA)
+
+thingB :: Int
+thingB = thingA + 1

--- a/tests/targets-that-must-fail/transitive-deps/LibC.hs
+++ b/tests/targets-that-must-fail/transitive-deps/LibC.hs
@@ -1,0 +1,6 @@
+module LibC (thingC) where
+
+import LibB (thingB)
+
+thingC :: Int
+thingC = thingB * 2

--- a/tests/targets-that-must-fail/transitive-deps/LibD.hs
+++ b/tests/targets-that-must-fail/transitive-deps/LibD.hs
@@ -1,0 +1,7 @@
+module LibD (thingD) where
+
+import LibA (thingA)
+import LibB (thingB)
+
+thingD :: Int
+thingD = thingA + thingB

